Add support for jmx metrics and prometheus exporter

kempspo opened the pull request.

nineinchnick commented:
I guess this would replace #140 and #135. Please think about how these changes can be tested - if we're adding resource manifests for external services, we need to deploy them too in the CI to make sure everything works as expected.

kempspo commented:
Added the extra services to the CI, but not sure if that's the best way.

nineinchnick commented on test.sh:
```diff
@@ -74,6 +74,12 @@ openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 \
+helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
+helm repo update
+helm install prometheus-operator prometheus-community/kube-prometheus-stack -n "$NAMESPACE"
+kubectl wait --for=condition=Ready pods -l "release=prometheus-operator" -n "$NAMESPACE"
```
Thanks for adding this! This and changes in `test-values.yaml` will prove there are no syntax errors in templates, and we can deploy the chart. The only missing piece is a test in `charts/trino/templates/tests/`. The existing `test-connection` runs a query in Trino to verify the basic connectivity. I think we should add a new one, that would query Prometheus and verify the metrics are being collected.

nineinchnick replied:
Should we wait for the pods, or use `kubectl rollout status --watch`?

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -727,6 +727,54 @@ kafka:
+  pullPolicy: Always
+  port: 5556
+  configProperties: []
+  # configProperties -- JMX Config Properties is mounted to /etc/jmx-exporter/jmx-exporter-config.yaml
```
You have to use the fully qualified property name in the beginning of the comment.
```suggestion
  # jmxExporter.configProperties -- JMX Config Properties is mounted to /etc/jmx-exporter/jmx-exporter-config.yaml
```

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -727,6 +727,54 @@ kafka:
+  registryPort: 9080
+  serverPort: 9081
+
+jmxExporter:
```
How about making this a `exporter` subsection in `jmx`?

nineinchnick replied:
OTOH Trino also supports OpenMetrics, so the JMX Exporter might not be always necessary: https://github.com/airlift/airlift/blob/master/openmetrics/src/main/java/io/airlift/openmetrics/MetricsConfig.java#L34

OpenMetrics are not documented yet.

This is a nitpick comment, feel free to ignore.

nineinchnick commented on test.sh:
```diff
@@ -74,6 +74,12 @@ openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 \
 
+echo 1>&2 "Installing Prometheus Operator"
+helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
+helm repo update
```
This takes a few seconds, which matters when tests are run repeatedly. Can we only run this when necessary?

kempspo replied:
This makes sense to me. What are the cases you think it should or shouldn't run?

nineinchnick replied:
Is it even necessary? I think `helm repo add` fetches the repo contents. I checked that `helm search repo prometheus` doesn't return any results after removing the repo, and does return expected results after adding it back with `helm repo add`, but without calling `helm repo update`.

kempspo replied:
I think `helm repo update` would only be necessary for local development where people could potentially have outdated versions of the helm repo. There might be a better way of notifying instead of forcing the update.

nineinchnick replied:
If we should be concerned with updates, shouldn't we deploy a specific version of the prometheus stack?

nineinchnick commented on test.sh:
```diff
@@ -74,6 +74,12 @@ openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 \
+echo 1>&2 "Installing Prometheus Operator"
+helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
+helm repo update
+helm install prometheus-operator prometheus-community/kube-prometheus-stack -n "$NAMESPACE"
```
Will this fail when running tests repeatedly without cleanup?

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -727,6 +727,54 @@ kafka:
+  #         type: UNTYPED
+
+podMonitor:
+  # - podMonitor is a resource from prometheus-operator:
+  #   https://github.com/prometheus-operator/prometheus-operator
```
```suggestion
  # podMonitor.enabled -- set to true to create resources for the [prometheus-operator](https://github.com/prometheus-operator/prometheus-operator).
```

nineinchnick replied:
Sorry I missed it initially, but please start sentences with a capital letter, so it'll be `Set to true...`

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -727,6 +727,54 @@ kafka:
+  # - podMonitor is a resource from prometheus-operator:
+  #   https://github.com/prometheus-operator/prometheus-operator
+  enabled: false
+  # labels for podMonitor, so that Prometheus can select it
```
To have these render in the README, start every comment with `--` - no need to repeat the fully qualified property name in one line comments, but the double dash is requried.

nineinchnick commented on charts/trino/values.yaml:
````diff
@@ -727,6 +727,54 @@ kafka:
   # ```
 
+jmx:
+  # https://trino.io/docs/current/admin/jmx.html
````
```suggestion
  # -- set to true to enable the RMI server to expose Trino's [JMX metrics](https://trino.io/docs/current/admin/jmx.html).
```

nineinchnick reviewed (approved):
Looking good, thanks for adding tests! If you'll address all comments, please squash commits, keeping logical changes separate:
* bumping version
* adding new properties

mosabua commented:
I dont have time to really review but just wanted to mention that this should all be optional and disabled by default.

nineinchnick commented on charts/trino/templates/tests/test-jmx.yaml:
```diff
@@ -0,0 +1,32 @@
+      image: {{ include "trino.image" . }}
+      command: ["/bin/bash", "-c"]
+      args:
+      - sleep 60
```
This should be a loop with a timeout. It might be better to write the whole test in Python, and run it using the `python:3-slim` image. We should not rely on the fact that Python is bundled in the Trino image.

I'm not requesting changes here, we can do that in a follow-up PR.

kempspo replied:
Yeah this makes much more sense to me.

nineinchnick commented on charts/trino/values.yaml:
```diff
@@ -727,6 +727,53 @@ kafka:
+  selector:
+    prometheus: kube-prometheus
+  # serviceMonitor.path -- The serviceMonitor web endpoint path
+  path: /admin/metrics
```
Is this the JMX Exporter path? If yes, how this is configurable in the JMX Exporter? Doesn't it default to `/metrics`? I can't find any references in the Prometheus Operator docs.

kempspo replied:
Yes, you're correct. It's on the [CRD](https://github.com/prometheus-community/helm-charts/blob/b651995181664fef3be46ef52ab963fb7899a009/charts/kube-prometheus-stack/charts/crds/crds/crd-servicemonitors.yaml#L455-L460)

nineinchnick replied:
We only need to support what we test, so let's not worry about OpenMetrics for now.

nineinchnick commented on test.sh:
```diff
@@ -74,6 +103,25 @@ openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 \
 
+# only install prometheus helm chart for these tests
+PROMETHEUS_TESTS=(complete_values)
+for test_name in "${PROMETHEUS_TESTS[@]}"; do
```
Looping here will install the Prometheus Operator multiple times, if we add more tests in the future. This loop seems unnecessary.

You could also simplify searching through the array with:
```
if printf '%s\0' "${TEST_NAMES[@]}" | grep -qwz complete_values
```

nineinchnick commented on test.sh:
```diff
@@ -74,6 +103,25 @@ openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 \
+        fi
+    done
+    if $found; then
+        uninstall_helm_repo "prometheus-operator"
```
This should be done during cleanup after tests.

nineinchnick commented on test.sh:
```diff
@@ -16,6 +16,35 @@ function join_by {
+    chart=$2
+    version=$3
+    echo 1>&2 "Installing $repo_name"
+    helm install "$repo_name" "$chart" -n "$NAMESPACE" --version "$version"
```
Maybe using `helm upgrade --install` instead would allow running this without checking if it's already installed.

nineinchnick commented on test.sh:
```diff
@@ -84,7 +132,29 @@ for test_name in "${TEST_NAMES[@]}"; do
+            # Check if there are any failed containers
+            if [ -n "$containers" ]; then
+                # Iterate through each failed container
+                for container in $containers; do
```
This is getting complex. Shellcheck will complain that `$containers` is not quoted, and it should be an array. Maybe we can just print all container logs with `kubectl logs --all-containers=true`, and make sure those who are successful don't print unnecessary output?

kempspo replied:
This option completely makes sense to me. I just realized `test-connection.yaml` currently logs a fair bit of stuff even on success.

nineinchnick replied:
You can remove `--network-logging=BODY` from it, I was too eager to add it (never needed it). Maybe add `--no-progress`, I'm not sure if it automatically detects there's no TTY, but explicit is better than implicit anyway.

nineinchnick commented on test.sh:
```diff
@@ -84,7 +132,29 @@ for test_name in "${TEST_NAMES[@]}"; do
     if ! time ct install "${CT_ARGS[@]}" --helm-extra-set-args "$HELM_EXTRA_SET_ARGS ${testCases[$test_name]}"; then
         echo 1>&2 "❌ Test $test_name failed"
         echo 1>&2 "Test logs:"
-        kubectl --namespace "$NAMESPACE" logs --tail=-1 --selector app.kubernetes.io/component=test
```
You don't really need to filter out successful pods/containers. Would be enough to add `--all-containers=true` here and avoid all the loops.

nineinchnick commented on test.sh:
```diff
@@ -16,6 +16,35 @@ function join_by {
+    repo_name=$1
+    repo_url=$2
+    echo 1>&2 "Adding Helm Repo $repo_name"
+    helm repo add "$repo_name" "$repo_url"
```
6 extra lines to wrap a single command seems an overkill, especially since this function is called exactly once.

kempspo replied:
Yeah definitely went overboard for some of these.

nineinchnick reviewed (commented):
Almost there

kempspo commented:
I realized I was trying to modify the `test.sh` to work on repeated runs when cleanup is skipped. But, two things would stop it even before it gets to the Prometheus items 
1. `kubectl create namespace` would fail if namespace exists
2. `kubectl create secret` would fail if secret exists in namespace
I can do a follow up PR to fix these.

kempspo commented on test.sh:
```diff
@@ -16,6 +16,35 @@ function join_by {
 
+function uninstall_helm_repo() {
+    repo_name=$1
+    # clusterrole being pesky here so just uninstall in all namespaces
```
I made a few changes that should simplify `test.sh` a lot.

While developing, I kept thinking that I had to do everything in the same namespace, but in reality, people would be deploying their monitoring stack on a different namespace.

I switched it to install on default, which now lets us do `helm upgrade --install` without any issues with the Cluster level resources when trying to reinstall again if cleanup was skipped.

That said, there are hanging CRDs that don't get removed by helm uninstall and must be manually deleted by `kubectl delete crd`

nineinchnick replied:
This is supposed to only work for testing, not on production clusters. But cleanup is important in case someone accidentally runs this on a cluster that's not ephemeral, like `kind`.

nineinchnick commented:
> I realized I was trying to modify the `test.sh` to work on repeated runs when cleanup is skipped. 

By default, the namespace is random. If you specify a namespace, you're supposed to clean it up manually. But we certainly could ignore the existing namespace and the secret, since it does make sense to avoid using a random namespace when debugging issues locally. Thanks for being patient with all the issues here.

nineinchnick commented on test-values.yaml:
```diff
@@ -140,6 +140,29 @@ accessControl:
+  enabled: true
+  selector:
+    prometheus: default
+  path: /admin/metrics
```
Remove this

nineinchnick commented on charts/trino/templates/servicemonitor.yaml:
```diff
@@ -0,0 +1,21 @@
+      {{- include "trino.selectorLabels" . | nindent 6 }}
+  endpoints:
+    - port: jmx-exporter
+      path: {{ .Values.serviceMonitor.path }}
```
This is not present in the default values anymore, remove

nineinchnick reviewed (approved):
Two last comments, squash commits and I can merge this.

nineinchnick commented:
@kempspo I applied some minor changes and opened #183, preserving your authorship of commits. I hope you don't mind.

I rewrote the tests in Python. I think they didn't work, the test container might have completed successfully even if the commands inside failed. Took me a bit to get it right, that's why I know it works now. BTW I removed the `release` label from the service monitor and instead added `prometheus.prometheusSpec.serviceMonitorSelectorNilUsesHelmValues=false` when deploying the operator, so it can grab any service monitor.

nineinchnick commented:
BTW I could have pushed my changes here if you've opened the PR from any other branch than `main`, since it's protected.

kempspo commented:
> @kempspo I applied some minor changes and opened #183, preserving your authorship of commits. I hope you don't mind.
> 
> I rewrote the tests in Python. I think they didn't work, the test container might have completed successfully even if the commands inside failed. Took me a bit to get it right, that's why I know it works now. BTW I removed the `release` label from the service monitor and instead added `prometheus.prometheusSpec.serviceMonitorSelectorNilUsesHelmValues=false` when deploying the operator, so it can grab any service monitor.

No worries. Those changes make sense to me.

Yeah I realized just now that I've been working on `main` 😅
